### ISSUE TITLE
vmselect: ability to set connect / response timeout from storage nodes

### ISSUE THREAD

**Issue by wf1nder** [VictoriaMetrics/VictoriaMetrics#711]
When one or more storage nodes under vmselect became unavailable, or begins to respond slowly due to high load, every request to vmselect start executing slowly. Obviously, vmselect waiting for each storage endpoint to generate a responce.

My suggestion is add ability to specify connect and response timeouts from storage endpoints. In that case we can tune up connect timeouts to smaller value (<1s for example), and the temporary loss of some storage nodes, which has HA-pairs, which are still available, will no longer be a problem for users.

**Comment by valyala** [comment]
The commit 9793008734a95bb590ca98c8f536668e66051018 adds `-search.storageTimeout` command-line flag to `vmselect`, which limits query execution time on every `-storageNode`. Note that `-search.denyPartialResponse` flag must be disabled, otherwise errors will be returned instead of partial responses. It is expected that `-search.storageTimeout` is smaller than `-search.maxQueryDuration` - otherwise `timeout exceeded` error is returned even if a part of `vmstorage` nodes already return results to `vmselect`.

@wf1nder , try building cluster components from the commit 9793008734a95bb590ca98c8f536668e66051018 and verifying whether `-search.storageTimeout` command-line flag works as expected for your case.

**Comment by wf1nder** [comment]
Unfortunately this didn't work in my case.

I started vmselect, builded today from master, with couple of ```-storageNode``` targets, some of them do not responding right now (which I filtered by ```ip6tables -A INPUT -s <vmstorage_host> -j DROP```.  I set ```-search.storageTimeout``` to 1s, and didn't specify ```-search.maxQueryDuration```,  so it uses default value 30s.
As a result, vmselect responds to queries with varying success. Sometimes it responds, but with a delay of 1-10 seconds. At that moment it writes to log following warnings:
```
2020-09-17T08:48:44.758Z        warn    app/vmselect/main.go:150        remoteAddr: "[<addr>]:60090"; cannot handle more than 16 concurrent search requests during 10s; possible solutions: increase `-search.maxQueueDuration`, increase `-search.maxConcurrentRequests`, increase server capacity
```

Sometimes after 10 seconds request to vmselect failing with following error in grafana:
```
remoteAddr: "[<addr>]:55104", X-Forwarded-For: "<addr>, <addr>, <addr>"; cannot handle more than 16 concurrent search requests during 10s; possible solutions: increase `-search.maxQueueDuration`, increase `-search.maxConcurrentRequests`, increase server capacity
```

And following warn in vmselect log:

```
2020-09-17T08:42:05.527Z        warn    app/vmselect/main.go:150        remoteAddr: "[<addr>]:46848", X-Forwarded-For: "<addr>, <addr>, <addr>"; cannot handle more than 16 concurrent search requests during 10s; possible solutions: increase `-search.maxQueueDuration`, increase `-search.maxConcurrentRequests`, increase server capacity
```

Also, a couple of times I've seen these errors:
```
2020-09-17T08:51:40.455Z        error   lib/netutil/tcplistener.go:69   temporary error when listening for TCP addr "[::]:8481": accept tcp [::]:8481: accept4: too many open files
```

When I removed ip6tables rule, and all ```-storageNode``` targets became available, it started to respond instantly and withour errors.

**Comment by valyala** [comment]
It looks like you hit the limit for concurrently executed queries, which can be adjusted with `-search.maxConcurrentRequests` command-line flag. If the limit is reached, then `vmselect` puts incoming requests in a wait queue until free slots are available for query processing. Requests may wait in the queue for up to `-search.maxQueueDuration`. After that `vmselect` logs the `cannot handle more than ... concurrent search requests` warning. Default value for `-search.maxQueueDuration` is 10 seconds.

Why requests may wait for 10 seconds in the queue if every request duration is limited by `-search.storageTimeout=1s`? Let's look at the following sequence:
* `-search.maxConcurrentRequests=16` concurrent requests hit `vmselect`. All of them wait for `-search.storageTimeout=1s` due to the slow `vmstorage` node.
* Additional requests arrived while the first 16 requests are waiting for response from slow `vmstorage` node. They are put in the wait queue. If these requests arrive at rate higher than `search.maxConcurrentRequests / search.storageTimeout`, i.e. 16 req/s for the case above, then the queue will constantly grow. This means that the majority of incoming requests will never reach `vmstorage` because the will be timed out in the wait queue.

How to deal with the issue for the given constant requests rate? Either increase `search.maxConcurrentRequests` or decrease `search.storageTimeout`.

The commit 69eb9783e6f5e7398faa1a5daad1765f9d5ed769 should partially address the issue by limiting max wait time in the queue to `min(search.maxQueryTimeout, search.storageTimeout)`, so now all the requests must be finished in `-search.storageTimeout`.

**Comment by valyala** [comment]
Note that too big values for `search.maxConcurrentRequests` may result in `vmstorage` overload and trashing. So this config must be adjusted carefully.

**Comment by wf1nder** [comment]
Sorry for the late reply.

Ok, I think I more or less understood this logic. I builded vmselect from fresh master, and checked it, varying settings ```-search.maxConcurrentRequests```, ```-search.storageTimeout``` and ```-search.maxQueueDuration```. Unfortunately, I was not able to get the desired behavior.

If I understand correctly, option ```-search.storageTimeout``` defines the whole timeout for request, starting from connect, and ending with getting reply from vmstorage. The problem is that we have graphs with heavy requests, which execution can usually take long time, 10+ seconds. So, decreasing value of ```-search.storageTimeout``` breask those graphs. And vice versa, increasing this setting means we are waiting too much time when one or move vmstorage nodes currently offline.

It may be worth adding the ability to set **connection timeout** separately the whole query execution timeout? Then we would be able to cut off earlier and not wait for offline nodes.

**Comment by wf1nder** [comment]
It OOM'ed :(

![Screenshot 2020-10-21 at 18 10 00](https://user-images.githubusercontent.com/4486762/96739697-bec68400-13c8-11eb-9477-5122206ec327.png)

Looks like some memory leak, because memory consumption is growing over time stepwise. This is also noticeable on bare metal servers (the arrow marks the moment of restart):

![Screenshot 2020-10-21 at 18 05 14](https://user-images.githubusercontent.com/4486762/96739331-5c6d8380-13c8-11eb-8e90-6853c14edbfe.png)

**Comment by wf1nder** [comment]
Another suggestion is to implement something like health checks on vmselect side to vmstorage nodes. On vmselect periodically (each ```n``` seconds) "ping" vmstorage nodes, and mark as "offline" ones, which is not responding, till next check. And exclude "offline" nodes when executing queries from clients.

**Comment by valyala** [comment]
@wf1nder , could build vmselect from the commit 85eecf5801d723c18631e5ac5e9bc563ba53bc59 and pass `-replicationFactor=N` command-line flag to it, where `N` equals to `-replicationFactor` used at `vminsert` nodes? This commit instructs `vmselect`to return response as soon as `vmstorage_nodes_count - replicationFactor + 1` vmstorage nodes successfully return responses. This number of responses from vmstorage nodes is enough for constructing full response. This should fix the original issue described in the first message above.

**Comment by valyala** [comment]
oops, there was a typo in the commit mentioned above, which resulted in a panic during vmselect startup. The typo has been fixed in 1dcb438c3bae6e5a36170103120e1030bc8e8385 .

**Comment by valyala** [comment]
There was a data race, which could result in panic if `-replicationFactor` command-line flag at `vmselect` is set to value higher than 1. It has been fixed in 25a57ced6c4b15ef817165f9a59d67c3377093a0 . @wf1nder , could you build `vmselect` from the latest commit in `cluster` branch (currently this is 2cc288c02376fca1b7ae9fe7d511c0e473c5f6dc ) and verify whether it properly handles queries when a part of vmstorage nodes are slow or unavailable?

**Comment by valyala** [comment]
FYI, `vmselect` starts supporting `-replicationFactor` command-line flag as described above starting from [v1.49.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by wf1nder** [comment]
I builded and checked vmselect from current master, and it seems works as you described: it doesn't dies when one of vmstorage shards became offline. And this is a big improvement!

But it is not ideal :)

If I understand correctly, now vmselect can wait for part of vmstorage shards answers, and construct response based on fastest responded vmstorage nodes.
But there might be situation, when some currently available nodes have no full set of data. For example, some vmstorage node was offline last couple of hours for some reason, and now became available. It responds now, but it have no data for some period.
As a result, vmselect can use it to construct respond, and not use some other vmstorage node, which have all set of data, but which responded a little slower.

In this case a good solution would be to use the answers of all currently available vmstorage nodes to construct responce to client.

Yes, I want to get rid of promxy :)

**Comment by wf1nder** [comment]
Ok, it is not the best example when one of vmstorage shard became offline, because vminsert can route writes to its neighbours, and there is no data loss. But there may be a situation when all shards in location became unavailable. So there may be gap of data in this location generally.

**Comment by valyala** [comment]
> But there may be a situation when all shards in location became unavailable. So there may be gap of data in this location generally.

Yes. But in this case gaps remain even if `vmselect` fetches data from all the `vmstorage` nodes.

**Comment by wf1nder** [comment]
But why? :)

Our desired architecture (simplified):

![vm](https://user-images.githubusercontent.com/4486762/101459793-beb71f00-3949-11eb-97e5-164d361ff674.png)


vmselect is enough 3 vmstorage shards to work and properly serve requests, because we can lost one of two VM clusters, and we need that the VM works in this case.
Let's imagine that cluster 2 was offline last couple of hours. All this time full set of metrics was delivering via cluster 1, so there is no any gaps on graphs.
Then cluster 2 back online. It doesn't have data for last hours, but it responds faster for some reason, so vmselect will construct responses based on data from it. All graphs will have gaps for last couple of hours despite the fact that cluster 1 still have full set of data without gaps.

Actually now we using promxy to workaround this issue:

![vm-with-promxy](https://user-images.githubusercontent.com/4486762/101459820-c8408700-3949-11eb-9dc7-2830ca981adc.png)


With enabled ```-search.denyPartialResponse``` on vmselect's. But it turns out not the most concise architecture with multiple vmselects, and some promxy-specific issues in addition.

**Comment by bogdanov1609** [comment]
ping? 
We are still facing this issue.

**Comment by valyala** [comment]
Unfortunately `vmselect` is unaware of the scheme outlined on the first image in the message https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-740472504 . `vmselect` assumes that all the `vmstorage` nodes are located in the same network in a single datacenter. This is needed for spending identical amounts of time when sending concurrent requests to all the `vmstorage` nodes and receiving responses from them. If these nodes are spread among multiple networks and/or datacenters, then connections from `vminsert` to `vmstorage` nodes could have different ping times and different network bandwidth (e.g. connections to `vmstorage` nodes in local datacenter usually have lower ping times and higher network bandwidth compared to connections to `vmstorage` nodes in remote datacenter). This may result to uneven response times and unneeded response delays.

`vmselect` nodes should know about network topology (e.g. about distinct `vmstorage` replication groups located in distinct datacenters) in order to support the scheme from the message https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-740472504 . I need to think more about such scheme support in VictoriaMetrics cluster.

**Comment by wf1nder** [comment]
Sorry to repeat myself, but it seems that everything could be solved by adding connection timeouts  into vmselect when accessing vmstorage, separate from the total timeouts for queries. Or, as a more complex solution, by adding periodic background health checks of vmstorage endpoints liveness, and ban those who do not respond, until they became evailable.

**Comment by wf1nder** [comment]
I think the ```-search.storageTimeout``` option doesn't work as expected, or its logic not obvious.
I create simple test stand: one vmselect (v1.56.0-cluster), which works with a single remote vmstorage. There is no any requests to this vmselect except for those single requests described below, so queues should not be clogged.
vmselect started with options:

```-enableTCP6 -search.disableCache -storageNode=<vmstorage_endpoint> -search.storageTimeout=1s```

And I querying it with request, which usually takes some time to execute.
Expectation: query must be failed right after 1 second.
Reality:

```
# time curl -s -X POST 'http://127.0.0.1:8481/select/0/prometheus/api/v1/query' -d '<some_heavy_request>'

{"status":"error","errorType":"422","error":"error when executing query=\"<some_heavy_request>" for (time=1616917615819, step=300000): error occured during search: cannot fetch query results from vmstorage nodes: cannot perform search on vmstorage <vmstorage_enddpoint>: cannot execute rpcName=\"search_v4\" on vmstorage \"<>" with timeout 30.000 seconds (elapsed 5.001 seconds); the timeout can be adjusted with `-search.maxQueryDuration` command-line flag: cannot read MetricBlock #12134: cannot read data with size 551: cannot read data in 0.025 seconds: cannot read data from the underlying reader: read tcp [<>]:55556->[<>]:8450: i/o timeout; read only 447 bytes"}
real	0m4.196s
user	0m0.006s
sys	0m0.004s
```

For some reason, execution of request is failing after more than 1 second, it tooks 4.2s. Repeating the query getting approximately the same results.
I experimented a bit with ```-search.storageTimeout``` option. When I set it to 5s, query fails after about 12s. On setting ```-search.storageTimeout``` to "10s", query fails after about 22s.

So looks like ```-search.storageTimeout``` option doesn't work as expected, it behaves unpredictably. And it's easy to reproduce.
The basic goal: to be able to set the maximum execution time for a query. To avoid waiting for storages that take too long to respond.

**Comment by wf1nder** [comment]
BTW, I found a workaround to implement connection timeouts between vmselect and vmstorage.
I put haproxy to container with vmselect. On container start, haproxy's config is generated with backend section for every remote vmstorage instances, which the vmselect needs to query. And vmselect requesting vmstorages through this haproxy, every vmstorage instance on separate port, instead of directly querying them.

Every haproxy's backend section is looks like this:

```
listen <instance-01>:8451
        bind                      :::9501
        server                    vmstorage <vmstorage-host-01>:8450 check on-marked-down shutdown-sessions inter 1s
```

Haproxy performing own tcp checks to every vmstorage. When one of vmstorage if stopped to respond, haproxy marking it as faulty, immediately resets all connections to correspond port, and starting to reset all new connections to this instance. So vmselect immediately stops to wait responses from this instance, and using responses only from alive vmstorage instances.

Thus, we able to put all vmstorage instances from all locations under one vmselect with haproxy. And we got adequate behavior in case of failures of individual vmstorage instances.
It would be cool to be able to set connection timeouts on vmselect itself, without the need to use haproxy.

**Comment by valyala** [comment]
It has been appeared that the `-search.storageTimeout` command-line flag has the same meaning as `-search.maxQueryDuration` command-line flag. That's why the `-search.storageTimeout` flag has been deleted in the commit ecfd6fe78d9892ef49378f966b4bb4ae77bd67b7 in favor to `-search.maxQueryDuration`.

> For some reason, execution of request is failing after more than 1 second, it tooks 4.2s. Repeating the query getting approximately the same results. I experimented a bit with -search.storageTimeout option. When I set it to 5s, query fails after about 12s. On setting -search.storageTimeout to "10s", query fails after about 22s.

This behavior has been improved in the commit ecfd6fe78d9892ef49378f966b4bb4ae77bd67b7 - now the maximum query duration shouldn't exceed 2 seconds compared to the value passed to `-search.maxQueryDuration`  command-line flag. Query duration can be also limited on a per-query basis with `timeout` query arg. For example, `/api/v1/query?query=foo&timeout=5s` should limit query duration to 5(+2) seconds.

**Comment by valyala** [comment]
The `-search.storageTimeout` command-line flag has been substituted by `-search.maxQueryDuration` starting from [v1.57.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by stillsnowy** [comment]
Hello. 

Any updates on this issue?

We facing the same problem described [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-701228964) and using promxy as frontend to workaround that issue

It seems that a flag to set the connection timeout rather than the entire query duration might solve this problem.

**Comment by sharninder** [comment]
We're also facing this issue. Setting up a promxy as a frontend is, at best, a workaround and, at worse, a hack. It'd be best to have health checks between vmselect and vmstorage but a timeout could also work in some cases.

**Comment by davidsome** [comment]
Hello.

Any updates on this issue?

We facing the same problem described

**Comment by valyala** [comment]
> Any updates on this issue?

We identified the root cause of the issue and working on the bug fix.

The root cause is in the hardcoded 5 seconds timeout for establishing TCP connections from `vmselect` to `vmstorage` nodes - see https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0abf46d66a97ed94ca51f438dff04f1547646b80/lib/netutil/tcpdialer.go#L20 . Usually `vmselect` nodes keep connection pools per each `vmstorage` node. Every connection is configured with [1 second keep-alive check](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0abf46d66a97ed94ca51f438dff04f1547646b80/lib/netutil/tcpdialer.go#L23). This means that the connection is automatically marked as broken in max 1 second after the `vmstorage` becomes unavailable. `vmselect` tries re-establishing new connections to `vmstorage` on new incoming requests in this case. It uses 5 second timeout for establishing such connections. Under normal conditions the operating system should almost immediately return error to `vmselect` when the given `vmstorage` node is unavailable. But sometimes the operating system may hang in `connect()` call to unavailable `vmstorage` node due to various reasons. For example, if DNS server cannot resolve the `vmstorage` hostname to IP address for long time. Or if the network between `vmselect` and `vmstorage` is improperly configured, so the operating system waits for `TCP connection rejected` response packet for too long. In this case the `connect()` call finishes with timeout error after the configured 5 seconds timeout.

We are going to implement the following solution for this issue:
1) To add `-vmstorageDialTimeout` command-line flag at `vmselect`, so it will be possible to change the default 5 seconds timeout to any desired value.
2) To add a background ping per every `vmstorage` node, which checks whether it is possible to connect to `vmstorage`. If the background ping detects that it is impossible to connect to a particular `vmstorage` node, then `vmselect` instantly skips querying such a node. Currently `vmselect` tries establishing connection to `vmstorage` node per each incoming query. This introduces the undesired delay per each query.

**Comment by valyala** [comment]
FYI, enhancements mentioned in [the comment above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-1160363187) have been included in [VictoriaMetrics v1.78.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by valyala** [comment]
Closing the feature request as done.

@wf1nder , feel free re-opening the feature request or creating a new one linked to this one if related issues still exist in VictoriaMetrics [v1.78.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by valyala** [comment]
FYI, the commit e1a2404db5e8e1248099750c6a4e45e90e2fa7be changes the logic behind `-replicationFactor` command-line flag at `vmselect`. Now it waits for responses from all the `vmstorage` nodes and returns full response when less than `-replicationFactor` `vmstorage` nodes are temporarily unavailable. The previous behavior - to skip responses from up to `-replicationFactor - 1` slowest vmstorage nodes - can be restored by passing `-search.skipSlowReplicas` command-line flag to `vmselect`.

This change was needed because many users were confused when `vmselect` could return incomplete responses when it runs with `-replicationFactor` identical to the `-replicationFactor` at `vminsert`. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1207) for details. The updated behavior for `-replicationFactor` command-line flag at `vmselect` should reduce the level of confusion in this case.

**Comment by valyala** [comment]
FYI, the behavior described [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-1627802973) is implemented in VictoriaMetrics since [v1.91.3](https://docs.victoriametrics.com/CHANGELOG.html#v1913)
